### ISSUE TITLE
What is preventing TF to use GPU when used in native windows?

### ISSUE THREAD

**Issue by eabase** [tensorflow/tensorflow#69750]
We keep getting told, repeatedly that TF no longer "supports" using GPU on versions `>2.10`. However, the reason why this is so, is nowhere to be found. 

**Can someone from the TF community please explain what is the problem with TF having GPU support for native windows environment?**

* https://github.com/tensorflow/tensorflow/issues/64881
* https://github.com/tensorflow/tensorflow/issues/63362
* https://www.tensorflow.org/install/pip#windows-native

Finally, since the wording is "supported" (and it was supported before), how can we go about implementing this support on our own? 

What are the python package requirements, and native windows C/C++/C# compiler requirements?

**Comment by sgkouzias** [comment]
@eabase according to [Simon_Au_Yong](https://discuss.tensorflow.org/t/help-needed-tensorflow-not-detecting-my-gpu/25015/9) in a relative discussion:

> Native Windows is challenging for the TF team to support given their limited resources. Hence them moving to WSL2 is understandable because it’s the best of both worlds, being both a Linux environment and having access to the GPU. The other option would be to boot direct from Linux from say an external SSD.
> 
> For professionals coming from other fields who have to work with neural nets Collab would be a quick option. Nothing to fuss about. And if you have the budget and need performance, [TPUs](https://www.tensorflow.org/guide/tpu) are a great choice.
> 
> Overall, the best environment to run TF with GPU support is on Linux x86-64 with the appropriate drivers installed.

I hope it helps!

**Comment by thiagojramos** [comment]
As you can see from the above comment, the only obstacle is the "Desire". Besides, their target audience doesn't seem to be Windows users (let alone home users who need to install this package because it's a requirement for something else).

**Comment by mihaimaruseac** [comment]
Up to 2022, I was in the team responsible for making TF work on all platforms. Windows, and especially GPU, was the one with the most breakages and also the one with the least expertise. Everyone was focused on Linux and only a team of ~3 was working on the other ecosystems.

Since then, none of those is still in TF and support is very reduced, even for Linux

**Comment by sgkouzias** [comment]
@mihaimaruseac I wonder why valuable TensorFlow team members left the team and why in general the number of members was so radically reduced. I really don't get it..

**Comment by eabase** [comment]
Hi Guys! 
Thank you so much for quick feedback. Not what I was hoping to hear, but nevertheless better than silence. 

@sgkouzias 
Hi Sotiris, 
I wasn't really looking for an answer that involves buying cloud server time. I have my own GPU, and want to put it to full use in all my environments, which is why WSL, is just not enough. I'm a huge fan of *nix, but at the end of the day, people wanna run in windows with whatever HW they have. SO it sounds very strange that "windows" is not supported. Everything else is running fine in windows, even using MSYS, Cygwin built tools etc, so I just don't see the issue here. Which is good, because it's also an incentive to get it working. 

@thiagojramos 
Hi Thiago, 
I would be careful to make assumptions on *target audience*. You have no idea what other fringe developers are trying to do and what they use. In my own case I am extremely *hybrid* in the sense that I try to use whatever native windows tools primarily, and then extend using *near* native tools as Cygwin, MSYS, and only WSL as a last resort, in which case I am not happy, because windows hides it's containers in an incompatible and non-portable way. If I have to use WSL, I  much rather prefer using a Virtualbox container where I have full control of everything while being able to backup and deployed on a different machine if necessary. I refuse to use Docker, as it completely shields developers from what is going on in the Linux environment, on the OS level, resulting in absurd communication and issues with app developers who doesn't understand basic *nix based OS principles. (Yeah, sure it is very useful for distributing quick tests and solutions, but there it ends, especially if you need to interact with hardware and bare metal.)

@mihaimaruseac 
Hi Mihai, Awesome!
**Can you point me in the right direction for the compilation process on windows using a nearly latest TF?** 
My first thought was that this should be possible using MSYS or MinGW64. 

---
`Rant Warning`:  
I then got *distracted* by Nvidia now saying to use their own Python repositories, installing absurd wrapper packages that then redirects the package repos to their own *pypi* servers. (JFC, who the heck was thinking this through? Totally senseless!)

**Comment by sgkouzias** [comment]
It looks like the issue is assigned to @Venkat6871 . @Venkat6871 could you provide some useful guidance?

**Comment by sgkouzias** [comment]
Professor [Alvaro Rodriguez](https://discuss.tensorflow.org/t/tensorflow-and-its-future/23869/6?u=sotiris.gkouzias) writes in a relevant discussion on the TensorFlow Forum:

"I have to say, that the drop of GPU support for windows, the lack of documentation and support for cpp, the lack of support and documentation for TensorFlow lite, the lack of TFrecord multi-platform standalone libraries… and so on is simply a strategy that will kill the library long term. Except for very niche projects in large companies.

Other platforms like PyTorch are investing in easy to use multi-platform solutions. If (or when) someone actually puts a solution powerful, stable, easy to access and easy to import and export to other platforms and languages. Private enthusiasts, researchers and academics will drop TensorFlow. And don’t forget that industries rely on specialists who learned in academia and come from research.

In a time of AI revolution, where the technology is more popular than ever, and is being added to literally everything. In my opinion, TensorFlow is neglecting everything outside Python-Linux, dropping an already lacking support for interoperability, and not investing in accessibility.

I’m saying this as a researcher and professor working in a computer science lab in an university. I write this just after investing almost 100 hours trying to simply build TensorFlow-cc to add some basic capabilities to a research project for the European Union, I failed. Also the absolute lack of recent information anywhere about TensorFlow-cc, and the responses I have seen to old threads lets me know most people gave up the same way I’m ready to tell my whole team to abandon TensorFlow and try other solutions.

I have seen others in my lab commenting similar concerns and frustrations. Many of our researchers are already moving away from TensorFlow and soon the whole department will follow.

For context. The Computer Science department is the largest department in my university, and serves the most important IT faculty in the Northwest of Spain.

For us ML is thriving. In addition to the Bachelor degree in Computer Science, where ML is more than present, being the most requested in the entire university. We opened a new one in Data Science, and are opening a new one in Artificial Intelligence. Next year we will be adding new classes and teachers to be able to serve the increasing number of students in two of the three Machine Learning subjects I teach… As far as I know, none of them will learn TensorFlow, none of them are using it in their personal projects and none of them will use it their degrees. They instead will be using Julia, Matlab, OpenCv , PyTorch, Scikit-Learn and other solutions.

Which leaves me to the industry sector. I worked also as a researcher in a public hospital in a project about diabetes, and in a private research center dedicated to laser and manufacturing. They all used TensorFlow, the same my laboratory did. I have been told they are all moving away from it, currently opting for a Scikit-Learn+OpenCv and PyTorch based approach. The reason is in one case the drop of GPU support for windows, and in the other a perceived drop of support combined with lack of interoperability.

The thing is, nobody moves away from a technology they spent years using and learning unless the technology fails them. And once you move away from something because of a problem, if you find a solution somewhere else, you will probably never return.

That is what is happening in TensorFlow. Google has intentionally dropped the ball with support, documentation, accessibility, ease of use, interoperability across languages and interoperability across platforms… so others will raise to the occasion.

Simply put, TensorFlow is becoming the Bing search engine with regards to AI"

**Comment by mihaimaruseac** [comment]
> @mihaimaruseac I am relatively a newbie in the world of TensorFlow but I wonder why valuable TensorFlow team members left the team and why in general the number of members was so radically reduced. I really don't get it..

Reorg during pandemic, shift of priorities, old manager and tech lead left, almost no-one who worked on TF pre-2.0 was still left in the team, new members with decision power but with no real knowledge of operating systems and compilers hired on time pressure, Peter principle, promo driven development, JAX, contracting support, Goodhart's law (in multiple instances). Read between the lines, but I cannot be more clear than this as I still have years of career left :)

> and so on is simply a strategy that will kill the library long term

JAX is what is the future now from Google. Or PyTorch. If you can use Keras 3 (which is not the default within TF -- or it should not be), you should be backend agnostic, but there's still a lot of work left to do to cover everything.

**Comment by mihaimaruseac** [comment]
> **Can you point me in the right direction for the compilation process on windows using a nearly latest TF?**

Unfortunately, most files that were there to support compiling on GPU on Windows are no longer in the repo, so there's not really a simple path forward.

**Comment by sgkouzias** [comment]
> > @mihaimaruseac I am relatively a newbie in the world of TensorFlow but I wonder why valuable TensorFlow team members left the team and why in general the number of members was so radically reduced. I really don't get it..
> 
> Reorg during pandemic, shift of priorities, old manager and tech lead left, almost no-one who worked on TF pre-2.0 was still left in the team, new members with decision power but with no real knowledge of operating systems and compilers hired on time pressure, Peter principle, promo driven development, JAX, contracting support, Goodhart's law (in multiple instances). Read between the lines, but I cannot be more clear than this as I still have years of career left :)
> 
> > and so on is simply a strategy that will kill the library long term
> 
> JAX is what is the future now from Google. Or PyTorch. If you can use Keras 3 (which is not the default within TF -- or it should not be), you should be backend agnostic, but there's still a lot of work left to do to cover everything.

@mihaimaruseac thank you so much for the insightful response. I am really grateful and do appreciate your efforts and generally the TensorFlow & Keras team (those who left the team and those who crafted and maintain the new, more inclusive and powerful Keras). I am a fan of the "_progressive disclosure of complexity_" core principle of Keras. If JAX is the future I will embrace it.

**Comment by eabase** [comment]
@mihaimaruseac  

>> Can you point me in the right direction for the compilation process on windows using a nearly latest TF?

> Unfortunately, most files that were there to support compiling on GPU on Windows are no longer in the repo, so there's not really a simple path forward.

I guess what I am saying, that the c/c++ used in the linux (WSL) compilation, should be straight forward to adjust to compile in native windows, unless they try to use native .NET/C#/Windows API calls. Forget all the stupid setup/build scripts, they only confuse everyone. This, of course, as long as there are no hard-coded libraries or other `*.so` files it relies on. (One should still be able to compile those, separately into dll's.) It's very weird that they don't have a git that shows when this was "removed" and *what* was "removed".

Anyway, thank you so much for providing the summary and clarity of the situation. Is anyone aware of some summary or methodology of translating legacy TF code into Torch or Keras?

**Comment by Venkat6871** [comment]
@learning-to-play

**Comment by mihaimaruseac** [comment]
> This, of course, as long as there are no hard-coded libraries or other `*.so` files it relies on.

This is the main issue. The Bazel files and macros that allowed compiling on GPU on Windows were excised. The macros assume in many places that they only build `*.so` files (no GPU support on Mac either).

Then, there's an issue that the old build scripts used MSVC + nvcc for CUDA compilation, but now everything is on clang, so the path on Windows + CUDA has not been tested at all.

**Comment by GatGit12** [comment]
Linking, relevant: https://github.com/tensorflow/tensorflow/issues/59918

**Comment by eabase** [comment]
@mihaimaruseac 
Do you have a good reference to the latest/best (or most complete) build instructions for using `clang + CUDA` on `WSL`?  

I can try to see if I can reproduce that build under MSYS. (I don't see why it couldn't be done.)

**Comment by mihaimaruseac** [comment]
I think running https://www.tensorflow.org/install/source under WSL should work

**Comment by GatGit12** [comment]
> @mihaimaruseac Do you have a good reference to the latest/best (or most complete) build instructions for using `clang + CUDA` on `WSL`?
> 
> I can try to see if I can reproduce that build under MSYS. (I don't see why it couldn't be done.)

Hello @eabase,
Have you made any progress and been able to build the latest TF with Clang + CUDA to get native GPU support  under Windows again?
Or is it just not possible due to the way TF is built (hardcoded libraries, complicated Bazel build scripts, ...)
I think it would be a great improvement if the latest TF would offer native GPU support on Windows again.

**Comment by ThomasHughesIV** [comment]
Does anyone know of an ETA on when this will be fixed?  This issue is blocking production.  I am unable to use Linux or containers and native support is a requirement.

**Comment by mihaimaruseac** [comment]
As discussed above, this is not getting fixed. There is no support for Windows + GPU in Tensorflow, this has been completely dropped in late 2022, early 2023.

**Comment by johnnkp** [comment]
I tried to compile 2.19 nightly under Windows CUDA this month. Finally stopped at `@local_xla//xla/stream_executor/cuda:delay_kernel_cuda` because of the following nvcc error:

```
external/com_google_absl\absl/functional/internal/any_invocable.h(521): error: enum "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...>::TargetType [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" has no member "kPointer"
              absl::enable_if_t<target_type == TargetType::kPointer, int> = 0>
                                                           ^
          detected during instantiation of class "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...> [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" at line 868

external/com_google_absl\absl/functional/internal/any_invocable.h(547): error: enum "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...>::TargetType [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" has no member "kCompatibleAnyInvocable"
                  target_type == TargetType::kCompatibleAnyInvocable, int> = 0>
                                             ^
          detected during instantiation of class "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...> [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" at line 868

external/com_google_absl\absl/functional/internal/any_invocable.h(548): error: invalid redeclaration of member function template "void absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...>::Initialize<target_type,QualDecayedTRef,F,<unnamed>>(F &&) [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" (declared at line 522)
    void Initialize(F&& f) {
         ^
          detected during instantiation of class "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...> [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" at line 868

external/com_google_absl\absl/functional/internal/any_invocable.h(560): error: enum "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...>::TargetType [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" has no member "kIncompatibleAnyInvocable"
                  target_type == TargetType::kIncompatibleAnyInvocable, int> = 0>
                                             ^
          detected during instantiation of class "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...> [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" at line 868

external/com_google_absl\absl/functional/internal/any_invocable.h(561): error: invalid redeclaration of member function template "void absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...>::Initialize<target_type,QualDecayedTRef,F,<unnamed>>(F &&) [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" (declared at line 522)
    void Initialize(F&& f) {
         ^
          detected during instantiation of class "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...> [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" at line 868

external/com_google_absl\absl/functional/internal/any_invocable.h(571): error: enum "absl::lts_20230802::internal_any_invocable::CoreImpl<SigIsNoexcept, ReturnType, P...>::TargetType [with SigIsNoexcept=false, ReturnType=ReturnType, P=<P...>]" has no member "kOther"
              typename = absl::enable_if_t<target_type == TargetType::kOther>>
```

Rewriting the code to remove `any_invocable` dependency is out of my ability, and it breaks Windows CUDA compatibility.

**Comment by ThomasHughesIV** [comment]
Since this product doesn't have basic native windows support, I've been successful in migrating to torch. Highly recommended

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
